### PR TITLE
ARROW-16467: [Python] Add helper function _exec_plan._filter_table to filter tables based on Expression

### PR DESCRIPTION
**Pull request by amol-**
The function is focused on Tables, as it will be the foundation for `Table.filter`, but as the extra work was minimal I added support for Dataset too.

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good. Code-wise just a small test suggestion.

I have one more general comment: for a filter operation of an in-memory table, I would personally expect my data to be kept in the same order (apart from the dropped rows) as the input. 
Does the exec plan already support tagging batches and reordering them when gathering into a table? I seem to recall that we do that for the Dataset to_table (but I don't know if this logic still lives in the scanner or has moved to exec plan)

**Comment by jorisvandenbossche** [comment]
> Does the exec plan already support tagging batches and reordering them when gathering into a table? I seem to recall that we do that for the Dataset to_table (but I don't know if this logic still lives in the scanner or has moved to exec plan)

cc @westonpace

**Comment by westonpace** [comment]
Yes, `to_table` supports this and it looks like that is what is being used here so it should work.

**Comment by jorisvandenbossche** [comment]
What I observed from trying out this branch is that it does not preserve:

```
In [18]: table1 = pa.table({'a': [1, 2, 3, 4], 'b': ['a'] * 4})

In [19]: table2 = pa.table({'a': [1, 2, 3, 4], 'b': ['b'] * 4})

In [20]: table = pa.concat_tables([table1, table2])

In [21]: ep._filter_table(table, pc.field('a') == 1)
Out[21]: 
pyarrow.Table
a: int64
b: string
----
a: [[1],[1]]
b: [["b"],["a"]]

In [22]: ep._filter_table(table, pc.field('a') == 1)
Out[22]: 
pyarrow.Table
a: int64
b: string
----
a: [[1],[1]]
b: [["a"],["b"]]
```

But the current cython wrappers of the ExecPlan are not using a "to_table" method, it is using `Table::FromRecordBatchReader`, with a record batch generator created using `MakeGeneratorReader`, which explicitly says it does not preserve order:

https://github.com/apache/arrow/blob/5a4b3db396f88898917620de50863b6d3a477a7a/cpp/src/arrow/compute/exec/exec_plan.h#L440-L447

I see that we currently use a "sink" node as final output node, but there is also a "table_sink" node. That might simplify the code a bit to get a Table (without manually creating a record batch reader and creating a table from that), but  don't think that would help with the ordering? (at least in the code I also don't see any ordering handling for this sink)

**Comment by jorisvandenbossche** [comment]
For the dataset scanner, I think the reordering logic lives in `AsyncScanner::ToTable()` / `AsyncTableAssemblyState`, which uses the fragment index to put the EnumeratedRecordBatch objects in the correct order in the resulting vector of batches.

**Comment by amol-** [comment]
> I see that we currently use a "sink" node as final output node, but there is also a "table_sink" node. That might simplify the code a bit to get a Table (without manually creating a record batch reader and creating a table from that), but don't think that would help with the ordering? (at least in the code I also don't see any ordering handling for this sink)

I tried and can confirm that using `table_sink` node doesn't help with preserving the order.

**Comment by amol-** [comment]
Given that preserving order in the proper way requires a non trivial amount of work from what I can see (not much of the required infrastructure is available outside of Datasets) I pushed a work-around in the current branch and I made https://issues.apache.org/jira/browse/ARROW-16518 to come up with a solution that works with threads too.

File: python/pyarrow/tests/test_exec_plan.py [hunk 1]
```diff
@@ -190,3 +191,35 @@ def test_table_join_keys_order():
...
+    result = ep._filter_table(
+        t, pc.field("b") > pc.scalar(30),
+        output_type=pa.Table if not use_datasets else ds.InMemoryDataset
+    )
+    if use_datasets:
+        result = result.to_table()
+    assert result == pa.table({
+        "a": [4, 5],
+        "b": [40, 50]
+    })
```
[the remark this answers is not in the document]
Comment (amol-) [reply python/pyarrow/tests/test_exec_plan.py:244]: 👍 done

**Comment by jorisvandenbossche** [comment]
Yes, that will indeed require some non-trivial work in the engine. Leaving it as is for now is good for me!

**Review by jorisvandenbossche — APPROVED** [review]

**Pull request closed without merging**